### ISSUE TITLE
RemoveEntriesOlderThan is not bounded

### ISSUE THREAD

**Issue by ramonsmits** [Particular/NServiceBus.NHibernate#143]
The current implementation both the client side queryover and server side delete statements are unbounded. If you have a very large number of items that will be deleted that that impacts locking and writing huge amounts of data in a single transaction to the journal log.

I better implementation is to do batched deletes in seperate transactions until there are no more items to be purged.

**Comment by ramonsmits** [comment]
After a chat with @DavidBoike we agreed that it would be hard to do in a database agnostic way.

We agreed on to test the affected row count and log a warning when a configurable arbitrary (6.000?) limit is passed. This means that if a customer is having more than 100 msg/s that he will be getting warnings to use a smaller cleanup interval.

https://github.com/Particular/NServiceBus.NHibernate/blob/develop/src/NServiceBus.NHibernate/Outbox/OutboxPersister.cs#L99


Also, the timer logic should make sure that the cleanup task cannot be overlapping. If should detect if the task is already running and if it is, log an error.

https://github.com/Particular/NServiceBus.NHibernate/blob/develop/src/NServiceBus.NHibernate/Outbox/NHibernateOutboxStorage.cs#L91

**Comment by ramonsmits** [comment]
To extend this, the timer logic should be auto sensing with a min and max period.

By default we should start with a minute.

- We should keep track of the last X (25?) deletion row counts and execution timestamps
- Sum those and divide that by an ideal (configurable) batch size (1.000?).
- Calculate the factor (batchsize/sum)
- Calculate the sleep time: period in milliseconds * factor
- min / max this
- Sleep :-)

In high volume environments we will automatically delete more often resulting in less likely to escalate to a page or table lock.

In low volume environments we will  maybe just run the cleanup every hour.

**Comment by ramonsmits** [comment]
Additional changes for consideration: https://groups.google.com/forum/#!topic/particularsoftware/CaZMSSti6BE

## TL;DR

- Improve cleanup logic

Also linked to #144 which is about index optimizations.

### Some background

> Our OutboxRecord table is stored in SQL Azure.  We have a high volume of traffic and we are seeing a large amount of database contention with selects, updates, and deletes.  The default installation of OutboxRecord is a table backed by a single database partition.  We're running a separate automation job to handle cleanup of the OutboxRecord table because the default delete logic is nearly always terminated by SQL Azure to avoid deadlocks of the other threads.  We have to delete from the table in smaller batches to avoid SQL terminating the process.

**Comment by ramonsmits** [comment]
I checked if this is still relevant and the logic is still the same:

https://github.com/Particular/NServiceBus.NHibernate/blob/0289c7540243afa506e9ff93188daa4c19ac12e9/src/NServiceBus.NHibernate/Outbox/OutboxPersister.cs#L134
